pkg/api/v1/router_firmware.go

```diff
@@ -0,0 +1,141 @@
+package serverservice
+
+import (
+	"github.com/gin-gonic/gin"
+	"github.com/volatiletech/null/v8"
+	"github.com/volatiletech/sqlboiler/v4/boil"
+	"github.com/volatiletech/sqlboiler/v4/queries/qm"
+
+	"go.hollow.sh/serverservice/internal/models"
+)
+
+func (r *Router) firmwareList(c *gin.Context) {
+	pager := parsePagination(c)
+
+	var params ComponentFirmwareVersionListParams
+	if err := c.ShouldBindQuery(&params); err != nil {
+		badRequestResponse(c, "invalid filter", err)
+		return
+	}
+
+	mods := params.queryMods()
+
+	dbFirmwares, err := models.ComponentFirmwareVersions(mods...).All(c.Request.Context(), r.DB)
+	if err != nil {
+		dbErrorResponse(c, err)
+		return
+	}
+
+	count, err := models.ComponentFirmwareVersions(mods...).Count(c.Request.Context(), r.DB)
+	if err != nil {
+		dbErrorResponse(c, err)
+		return
+	}
+
+	firmwares := []ComponentFirmwareVersion{}
```

Comment:
since you have the count of rows returned, you can preallocate `firmwares`
```
firmwares := make([]ComponentFirmwareVersion, count)
```

https://www.programming-books.io/essential/go/optimize-by-pre-allocating-slice-ee6501b596c9446da8b82e4ed35454a6

Reply:
Good tip, thanks! Fixed in [2f839d4](https://github.com/metal-toolbox/hollow-serverservice/pull/48/commits/2f839d4ca77b9c6ce7dfa3754e80584cec430264)